src/rpc/rawtransaction.cpp

```diff
@@ -229,29 +238,36 @@ UniValue gettxoutproof(const JSONRPCRequest& request)
        oneTxid = hash;
     }
 
-    LOCK(cs_main);
-
     CBlockIndex* pblockindex = nullptr;
-
     uint256 hashBlock;
-    if (!request.params[1].isNull())
     {
-        hashBlock = uint256S(request.params[1].get_str());
-        pblockindex = LookupBlockIndex(hashBlock);
-        if (!pblockindex) {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
-        }
-    } else {
-        // Loop through txids and try to find which block they're in. Exit loop once a block is found.
-        for (const auto& tx : setTxids) {
-            const Coin& coin = AccessByTxid(*pcoinsTip, tx);
-            if (!coin.IsSpent()) {
-                pblockindex = chainActive[coin.nHeight];
-                break;
+        LOCK(cs_main);
+
+        if (!request.params[1].isNull()) {
+            hashBlock = uint256S(request.params[1].get_str());
+            pblockindex = LookupBlockIndex(hashBlock);
+            if (!pblockindex) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+            }
+        } else {
+            // Loop through txids and try to find which block they're in. Exit loop once a block is found.
+            for (const auto& tx : setTxids) {
+                const Coin& coin = AccessByTxid(*pcoinsTip, tx);
+                if (!coin.IsSpent()) {
+                    pblockindex = chainActive[coin.nHeight];
+                    break;
+                }
             }
         }
     }
 
+    // Allow txindex to catch up if we need to query it and before we acquire cs_main.
+    if (g_txindex && !pblockindex) {
+        g_txindex->BlockUntilSyncedToCurrentChain();
```

Comment:
Also here - should report a different error of some kind if we're not in sync yet.

Reply:
The error here already is not that useful and does not even indicate if the txindex is disabled. I'd like to fix this in a follow-up PR.